### PR TITLE
Use gitpython to find a commit reference

### LINKED ISSUE #3842: Stable versions are broken
For instance: http://readthedocs.org/projects/docs/builds/6938938/

### PR DESCRIPTION
**Pull request by agjohnson**
Fixes #3842

### CODE REVIEW

**Comment by agjohnson** [comment]
Test failure is linting failure again.

**Comment by stsewd** [comment]
@agjohnson isn't better to just revert #3831 for now for the hotfix? I think probably others parts of the code depends on the original `exit_code`, and we don't want more surprises.

File: readthedocs/vcs_support/backends/git.py [hunk 1]
```diff
@@ -254,8 +256,13 @@ def find_ref(self, ref):
         return ref
 
     def ref_exists(self, ref):
-        code, _, _ = self.run('git', 'show-ref', ref, record_as_success=True)
-        return code == 0
+        try:
+            r = Repo(self.working_dir)
+            if r.commit(ref):
+                return True
+        except BadName:
```
Comment (davidfischer) [inline readthedocs/vcs_support/backends/git.py]: Perhaps this should be `except (BadName, ValueError):`. I believe `BadName` is raised if the string is invalid like a string that can't be a commit hash or isn't a branch. `ValueError` is raised if a commit is not found for some reason.

**Comment by stsewd** [comment]
Here are some others parts of code that depends on the original `exit_code`

```
readthedocs/vcs_support/backends/git.py|128 col 1| record_as_success=True,
readthedocs/doc_builder/environments.py|311 col 1| record_as_success=False, **kwargs)
readthedocs/vcs_support/backends/git.py|172 col 1| record_as_success=True,
readthedocs/doc_builder/environments.py|98 col 1| self.record_as_success = record_as_success
readthedocs/vcs_support/backends/hg.py|41 col 1| retcode, stdout = self.run('hg', 'branches', record_as_success=True)[:2]
readthedocs/vcs_support/backends/svn.py|68 col 1| % self.base_url, record_as_success=True)[:2]
readthedocs/vcs_support/backends/git.py|276 col 1| code, _, _ = self.run('git', 'show-ref', ref, record_as_success=True)
readthedocs/vcs_support/backends/bzr.py|48 col 1| retcode, stdout = self.run('bzr', 'tags', record_as_success=True)[:2]
readthedocs/doc_builder/environments.py|75 col 1| bin_path=None, description=None, record_as_success=False)
readthedocs/vcs_support/backends/hg.py|54 col 1| retcode, stdout = self.run('hg', 'tags', record_as_success=True)[:2]
readthedocs/vcs_support/backends/svn.py|37 col 1| retcode = self.run('svn', 'info', record_as_success=True)[0]
```

**Comment by agjohnson** [comment]
I don't think the problem is necessarily that we're using exit_code. I don't think we can remove dependency on that for some of these cases. I'd rather see if we can move these calls to using a library though, as relying on our own parsing and interpretation of exit code for everything is where these bugs come from.

To @humitos questions:

> how we will execute this inside Docker once we have the VCS code migrated to Docker?

Not sure yet. Monkey patching command execution to use our docker build command is a possibility. Not running in docker is another. Third would be application code that gets executed from the docker container

> how we will present these commands to the user in the build log?

Many of these commands aren't important, there are better ways we can show some of this data. If we can monkey patch, commands will execute as normal

**Pull request merged**